File: iteration2/CourseRegistrationSystem/src/JSONMethods.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.File;
import java.io.IOException;

public class JSONMethods {
    public boolean clearRequestedCourses(Student student) {

        try {
            // JSON file will be read
            File jsonFile = new File("src/Students/"+student.getStudentId()+".json");

            // Read JSON data as follows
            ObjectMapper objectMapper = new ObjectMapper();
            JsonNode rootNode = objectMapper.readTree(jsonFile);

            // Find the "requestedCourses" field
            JsonNode requestedCoursesNode = rootNode.get("requestedCourses");

            // If the "requestedCourses" field is not empty, delete the content
            if (requestedCoursesNode != null && requestedCoursesNode.isArray()) {
                ArrayNode requestedCoursesArray = (ArrayNode) requestedCoursesNode;
                requestedCoursesArray.removeAll();
            }

            // Write changes to JSON file
            objectMapper.writeValue(jsonFile, rootNode);

        } catch (IOException e) {
            e.printStackTrace();
        }
        student.getRequestedCourses().clear();
        return true;
    }

    public boolean addEnrolledStudent(Student student, Course course){
        try {
            // Öğrenciye ait JSON dosyasını oku
            File studentJsonFile = new File("src/Students/"+student.getStudentId()+".json");
            ObjectMapper objectMapper = new ObjectMapper();
            JsonNode studentNode = objectMapper.readTree(studentJsonFile);

            // "requestedCourses" alanını bul
            JsonNode enrolledCoursesNode = studentNode.get("enrolledCourses");

            // Eğer "requestedCourses" alanı boşsa oluştur
            if (enrolledCoursesNode == null || !enrolledCoursesNode.isArray()) {
                enrolledCoursesNode = objectMapper.createArrayNode();
                ((ObjectNode) studentNode).set("enrolledCourses", enrolledCoursesNode);
            }

            // Course bilgilerini yeni bir JSON nesnesine ekle
            ObjectNode newCourseNode = objectMapper.createObjectNode();
            newCourseNode.put("courseId", course.getCourseId());
            newCourseNode.put("courseName", course.getCourseName());
            newCourseNode.put("credit", course.getCredit());
            newCourseNode.put("prerequisite", course.hasPrerequisite());
            newCourseNode.put("prerequisiteLessonId", course.getPrerequisiteLessonId());

            // CourseSection bilgilerini ekleyin (bu kısmı kendi ihtiyaçlarınıza göre düzenleyebilirsiniz)
            ObjectNode courseSectionNode = objectMapper.createObjectNode();
            courseSectionNode.put("term", course.getCourseSection().getTerm());
            courseSectionNode.put("instructor", course.getCourseSection().getInstructor());
            courseSectionNode.put("enrollmentCapacity", course.getCourseSection().getEnrollmentCapacity());
            courseSectionNode.put("status", course.getCourseSection().getStatus());

            newCourseNode.set("courseSection", courseSectionNode);

            // "requestedCourses" alanına yeni kursu ekle
            ((ArrayNode) enrolledCoursesNode).add(newCourseNode);

            // Değişiklikleri JSON dosyasına kaydet
            objectMapper.writeValue(studentJsonFile, studentNode);

        } catch (IOException e) {
            e.printStackTrace();
        }
        return true;
    }
}

File: iteration3/JsonMethods.py
import json

class JSONMethods:
    def clear_requested_courses(student):
        try:
            file_path = f"src/Students/{student.get_student_id()}.json"
            with open(file_path, 'r+') as json_file:
                data = json.load(json_file)
                data['requestedCourses'] = []
                json_file.seek(0)
                json.dump(data, json_file, indent=4)
                json_file.truncate()
            student.clear_requested_courses()
            return True
        except Exception as e:
            print(e)
            return False

    def add_enrolled_student(student, course):
        try:
            file_path = f"src/Students/{student.get_student_id()}.json"
            with open(file_path, 'r+') as json_file:
                data = json.load(json_file)
                enrolled_courses = data.get('enrolledCourses', [])
                new_course = {
                    "courseId": course.get_course_id(),
                    "courseName": course.get_course_name(),
                    "credit": course.get_credit(),
                    "prerequisite": course.has_prerequisite(),
                    "prerequisiteLessonId": course.get_prerequisite_lesson_id(),
                    "courseSection": {
                        "term": course.get_course_section().get_term(),
                        "day": course.get_course_section().get_day(),
                        "hour": course.get_course_section().get_hour(),
                        "status": course.get_course_section().get_status(),
                        "semester": course.get_course_section().get_semester(),
                        "instructor": course.get_course_section().get_instructor(),
                        "enrollmentCapacity": course.get_course_section().get_enrollment_capacity(),
                        "status": course.get_course_section().get_status()
                    }
                }
                enrolled_courses.append(new_course)
                data['enrolledCourses'] = enrolled_courses
                json_file.seek(0)
                json.dump(data, json_file, indent=4)
                json_file.truncate()
            return True
        except Exception as e:
            print(e)
            return False

    def add_request_course(course, student):
        try:
            file_path = f"src/Students/{student.get_student_id()}.json"
            with open(file_path, 'r+') as json_file:
                data = json.load(json_file)
                requested_courses = data.get('requestedCourses', [])
                new_course = {
                    "courseId": course.get_course_id(),
                    "courseName": course.get_course_name(),
                    "credit": course.get_credit(),
                    "prerequisite": course.has_prerequisite(),
                    "prerequisiteLessonId": course.get_prerequisite_lesson_id(),
                    "courseSection": {
                        "term": course.get_course_section().get_term(),
                        "day": course.get_course_section().get_day(),
                        "hour": course.get_course_section().get_hour(),
                        "status": course.get_course_section().get_status(),
                        "semester": course.get_course_section().get_semester(),
                        "instructor": course.get_course_section().get_instructor(),
                        "enrollmentCapacity": course.get_course_section().get_enrollment_capacity() - 1,
                        "status": course.get_course_section().get_status()
                    }
                }
                requested_courses.append(new_course)
                data['requestedCourses'] = requested_courses
                json_file.seek(0)
                json.dump(data, json_file, indent=4)
                json_file.truncate()
        except Exception as e:
            print(e)

    def update_enrollment_capacity(course_id, new_capacity):
        try:
            file_path = "src/course.json"
            with open(file_path, 'r+') as json_file:
                data = json.load(json_file)
                for course in data:
                    if course['courseId'] == course_id:
                        course['courseSection']['enrollmentCapacity'] = new_capacity
                        json_file.seek(0)
                        json.dump(data, json_file, indent=4)
                        json_file.truncate()
                        return True
                return False
        except Exception as e:
            print(e)
            return False

    def update_project_assistant(student, project_assistant):
        try:
            file_path = f"src/Students/{student.get_student_id()}.json"
            with open(file_path, 'r+') as json_file:
                data = json.load(json_file)
                data['projectAssistant'] = project_assistant
                json_file.seek(0)
                json.dump(data, json_file, indent=4)
                json_file.truncate()
        except Exception as e:
            print(e)
